File: bishopTests/src/bishopTests/LineIndexerTest.java
package bishopTests;

import java.util.HashSet;
import java.util.Set;

import org.junit.Assert;

import org.junit.Test;

import bishop.base.BitBoardCombinator;
import bishop.base.CrossDirection;
import bishop.base.LineIndexer;
import bishop.base.Square;

public class LineIndexerTest {

	@Test
	public void testIndices() {
		final Set<Integer> indices = new HashSet<Integer>();
		
		for (int direction = CrossDirection.FIRST; direction < CrossDirection.LAST; direction++) {
			for (int square = Square.FIRST; square < Square.LAST; square++) {
				final long mask = LineIndexer.calculateDirectionMask(direction, square);
				
				for (BitBoardCombinator combinator = new BitBoardCombinator(mask); combinator.hasNextCombination(); ) {
					final long combination = combinator.getNextCombination();
					final int index = LineIndexer.getLineIndex(direction, square, combination);
					
					if (!indices.add(index))
						Assert.fail("Dupplicate combination: direction " + direction + ", square " + square + ", combination " + combination);
				}
			}
		}
	}
}

File: native/tests/bishop/line_indexer_test.cpp


#include "../../bishop/base/line_indexer.h"
#include "../../bishop/base/bit_board_combinator.h"
#include "../test_base/test_base.h"

#include <set>

using namespace bishop::base;


TEST(testIndices) {
	std::set<LineIndexer::IndexType> indices;
	
	for (CrossDirection::Type direction = CrossDirection::FIRST; direction < CrossDirection::LAST; direction++) {
		for (Square::Type square = Square::FIRST; square < Square::LAST; square++) {
			const BitBoard::Type mask = LineIndexer::calculateDirectionMask(direction, square);
			
			for (BitBoardCombinator combinator(mask); combinator.hasNextCombination(); ) {
				const BitBoard::Type combination = combinator.getNextCombination();
				const LineIndexer::IndexType index = LineIndexer::getLineIndex(direction, square, combination);
				assertTrue (index >= 0 && index < LineIndexer::LAST_INDEX);
				
				const bool dupplicate = !indices.insert(index).second;
				assertFalse (dupplicate);
			}
		}
	}
}

TEST_RUNNER_BEGIN
	RUN_TEST(testIndices);
TEST_RUNNER_END

